app/src/main/java/ca/marshallasch/veil/controllers/RightMeshController.java

```diff
@@ -0,0 +1,290 @@
+                    Toast.makeText(serviceContext, status, Toast.LENGTH_SHORT).show();
+                    return;
+                }
+                Log.d("MESH", "initilized");
+
+            case RESUME:  // over the mesh!
+
+                meshActive = true;
+                break;
+            case FAILURE:  // Mesh connection unavailable,
+            case DISABLED: // time for Plan B.
+                break;
+        }
+    }
+
+    /**
+     * Handles peer update events from the mesh - maintains a list of peers and updates the display.
+     *
+     * @param e event object from mesh
+     */
+    private void handlePeerChanged(MeshManager.RightMeshEvent e) {
+
+        // Update peer list.
+        MeshManager.PeerChangedEvent event = (MeshManager.PeerChangedEvent) e;
+        if (event.state != REMOVED) {
+
+            if (event.peerUuid.equals(meshManager.getUuid())) {
+                Log.d("FOUND", "found loopback: " + event.peerUuid);
+                return;
+            }
+
+            Log.d("FOUND", "found user: " + event.peerUuid);
+
+
+
+            Sync.HashData hashData = dataStore.getDataStore();
+            Sync.MappingMessage mappingMessage =  dataStore.getDatabase();
+
+            // send messages to the peer.
+            try {
+
+                Sync.Message message = Sync.Message.newBuilder()
+                        .setType(Sync.SyncMessageType.HASH_DATA)
+                        .setData(hashData)
+                        .build();
+                meshManager.sendDataReliable(event.peerUuid, DATA_PORT, message.toByteArray());
+
+                message = Sync.Message.newBuilder()
+                        .setType(Sync.SyncMessageType.MAPPING_MESSAGE)
+                        .setMapping(mappingMessage)
+                        .build();
+                meshManager.sendDataReliable(event.peerUuid, DATA_PORT, message.toByteArray());
+            }
+            catch (RightMeshException e1) {
+                e1.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Requests data from connected peers over the mesh and sends a broadcast notifying that
+     * data has been updated.
+     */
+    public void manualRefresh(){
+        try {
+            MeshManager manager = this.meshManager;
+            Set<MeshId> peers = manager.getPeers(this.DATA_PORT);
+
+            Sync.Message dataRequest = Sync.Message.newBuilder().setType(Sync.SyncMessageType.REQUEST_DATA).build();
+
+            // request an update from everyone
+            for (MeshId peer: peers) {
+
+                // do not ask myself for info
+                if (peer.equals(manager.getUuid())) {
+                    continue;
+                }
+                manager.sendDataReliable(peer, this.DATA_PORT, dataRequest.toByteArray());
+            }
+
+        }
+        catch (RightMeshException e) {
+            e.printStackTrace();
+        }
+        Intent intent = new Intent(RightMeshController.NEW_DATA_BROADCAST);
```

Comment:
This should not be broadcast here because the new data has not been received yet.

Reply:
This broadcast notifies the Receiver in FragmentDiscoverForums to pull for new data from the data store. Where would you suggest to move the broadcast?

Reply:
It will cause the `FragmentDiscoverForums` to refresh the data but it wont actually have any new data yet because it there has not been a response yet from the other devices.